### ISSUE TITLE
[5pt] graph-prototype:  Implementation of Basic Python Bindings Using pybind11

### ISSUE THREAD

**Issue by ivan-cukic** [fair-acc/gnuradio4#120]
The purpose of this issue is to introduce basic Python bindings using pybind11 for a simple graph. This graph is to comprise two sources (src#1: sine, src#2: static offset), an adder, a multiplier, and sinks. The project is split into three main tasks:

 1. The creation and connection of the graph from a Python script, followed by its output into a GRC YAML file.
 1. Reading the graph structure from the generated GRC file, followed by its execution.
 1. Comparing the outputs of tasks 1 and 2, ensuring they are identical.

Notably, the creation of a dedicated Python-only block/work signature is beyond the scope of this issue, which will be handled separately after the graph-prototype becomes the official GNU Radio 4.0 (GR 4.0).

Future extensions beyond this basic example and further work will be conducted by @mormj et al. once the graph-prototype is incorporated into the official GR repository.

I'd like to point out a critical point to bear in mind during this implementation is the instantiation of typed block templates. As Python lacks an analogous concept to C++ templates, which our blocks/processing nodes extensively use, these may need to be instantiated once before the pybind11 bindings can be generated. I think it would be useful if the pybind11 generator function accepts these template parameters and handles the template instantiation for the user.

**Comment by japm48** [comment]
Just a suggestion:
Have you considered nanobind instead of pybind11? It was created by the same person and my understanding is that it is quite better in many aspects, especially not having to depend on LTO. The only downside being that it does not support multiple inheritance, which I don't know if the redesigned architecture depends on or not.

**Comment by RalphSteinhagen** [comment]
@japm48, thanks for your interest and for bringing this up.

For GR 4.0 we do not intend to use (rather avoid) multiple inheritance because of performance issues and implications w.r.t. refactoring, being pinned-down to a niche, and long-term maintenance.

This may be a good time and place to discuss and evaluate the various options. We discussed this in a smaller setting within the [GR Architecture Group](https://matrix.to/#/#architecture:gnuradio.org). Please feel free and invited to join that channel as well. @mormj, @dkozel, @jsallay, @marcusmueller, @argilo, and @willcode please chime in for context

The primary issue is that there is a multitude of options, for example:
 * [cpython-ctype](https://github.com/python/cpython) -- *the bare-knuckle option*
 * [pypy-c](https://github.com/mozillazg/pypy) 
 * [pyind11](https://github.com/pybind/pybind11) 
 * [nanobind](https://github.com/wjakob/nanobind) -- *known for it's efficiency*
 * [cppyy](https://github.com/wlav/cppyy)  -- *probably the fasted option*
 * *<add your other favourite options here ...>*

All with their pros and cons. Our current API is designed to support runtime and C-Style/Python wrapping. At GSI/FAIR, we use Python primarily for prototyping and AI systems, we do not have strong opinions on the chosen option (other than that it should also minimise long-term maintenance). We would use what fits with the use case of most GNU Radio users. 

And herein lies the problem: there is a large 10k+ user community that is not very verbose on how they use GNU Radio and/or their application. Some are academic, some private enthusiasts, ... but also many important industrial use cases often shrouded in mystery and proprietary code/applications (*N.B., which is OK*). We (and I guess only a few) know what the most relevant, targeted, and sustainable use cases are. My best guesses are, using Python to
 1. graphically create, store, load, and interact with different flow-graphs using the GNU-Radio Companion (being written in Python) and it's `*.grc` yaml file-format.
 2. write new blocks functions that use the Python-Syntax (scripting) and/or interact with other Python-only libraries
 3. boot-strap GR flow-graphs w/o graphical user-interface using Python-only
 4. re-write custom/the entire GR block structure using Python from scratch ...?
 5. *<add your other favourite options here ...>*

The first two options are lightweight and require minimal to no Python wrapping. For example, the first option would primarily only wrap the existing C++-API for storing/loading of `*.grc` files, a bit of the scheduler-API, and John's PMT library so that GRC could interact with it. This would be a very limited sub-set. The second option could be implemented by invoking a local CPython instance -- or other Python interpreter, for that matter -- from the work function and thus avoiding the [GIL](https://realpython.com/python-gil/). This is a biased view, and since I am aware of this, I'd instead refer to those power Python users, developers, and maintainers... such as yourself as I gather.

I hope the above helps to put some of the questions into context and explain why we haven't written an official Python binding (yet).

Your input will help us understand the context better and guide our decision on implementing an official Python binding for GNU Radio 4.0.

**Comment by willcode** [comment]
If we don't worry about the performance of Python-only blocks, then the GIL isn't a huge driver. Numpy releases the GIL for (some?) routines. We don't currently `gil_scoped_release` in the bindings, and probably should. In almost all cases, this would be safe, since Python objects are not accessed by the wrapped C++ code. We do `gil_scoped_acquire` for calls from C++ to Python blocks.

cppyy says it uses LLVM, and we probably don't want to make that a manditory dependency. Nanobind sounds like it could be useful.

**Comment by jsallay** [comment]
I'll also note that nanobind uses a more limited set of Python known as the
Stable ABI. Starting with Python 3.12, anything built with nanobind will be
compatible with all future versions of Python without requiring a recompile.

**Comment by RalphSteinhagen** [comment]
> [..] we probably don't want to make that a mandatory dependency [..]

I am all for minimising dependencies. It's all about added value vs. the costs/size/maintenance/... burden they add. GCC and LLVM (two packages, ~140 MB, similar to Python+NumPy) are quite mature, dependable, and would allow also -- if we choose to -- add JIT-C++ capabilities to GR 4.0. The latter may be useful to provide compile-time optimised Sub]Graphs, GPU and/or FPGA support to a broader user-base.

In a way the GRC is already having a soft dependency on GCC/LLVM and via Python also a strong dependency on the stdc++ libs. Still, this would require a careful design and integration to keep this manageable...

**Comment by RalphSteinhagen** [comment]
The core question is: what do most people want to do with GR 4.0 using Python? 

I listed the four cases above but I guess
 a) that there are plenty more, and
 b) that some of them are quite niche.
 
 For me, it's hard to design a Python binding if the target isn't well-defined or fluent, depending on who is asked. :thinking:

**Comment by willcode** [comment]
If people can do the same thing they do now with Python bindings and blocks, but better, faster, stronger, that would be a good start for transition.

**Comment by RalphSteinhagen** [comment]
@willcode4 the core of our Python integration strategy in GR 4.0 is finding the 'right tool for the job.

The 80/20 rule, as you aptly mentioned, becomes crucial in determining where Python's inclusion offers substantial value without overcomplicating the system.
For me, the hard-to-answer question is: what do people _actually_ want to do with Python that cannot be done otherwise? The right balance is key - offering enough Python integration to empower our users without burdening the system with unnecessary complexities.

Do the following two use cases cover the 80% sufficiently or are there important ones missing:
 1. **GNU Radio Companion** (GRC): covers a significant portion of the workflow - such as
    * creating, and modifying of flow-graph structures,
    * loading/saving of *.grc files, 
    * executing flow-graphs and changing settings during runtime, and
    * graphically representing the signals (N.B. the non-graphical in/outs would probably go via network (0MQ, HTTP, ...) or pipes.
 2. **`PythonBlock`** (encapsulating CPython in a `Block<T>`): can be adapted for multiple stream and message inputs/outputs and would allow the ability to integrate various Python libraries (like Numpy, SciPy, Pandas, OpenCV, TensorFlow, PyTorch, Theano, etc.).

Do these capabilities cover the majority of Python-related requirements within our community? Are there specific use cases or workflows where Python offers a unique advantage that we have not yet encapsulated within GRC and the `PythonBlock`?

This is a genuine design/architecture question ...

I look forward to hearing more thoughts and insights on this. Notably, **please chime in if there are some core use cases that are not covered**.

Maybe someone has a crystal ball for this type of problems. :smile:

**Comment by willcode** [comment]
Use of generated Python top block classes in custom Python programs.

Ability to write a Python GNU Radio based program from scratch (often based on output from GRC).
